### ISSUE TITLE
Implement Compound collider bodies

### ISSUE THREAD

**Issue by nevernotsean** [react-spring/use-cannon#29]
From the example here: http://schteppe.github.io/cannon.js/demos/compound.html

> A Compound shape is a shape built out of other shapes called child-shapes.
> You can see it as a holder of a group of other shapes.
> Use the compound shape to build rigid bodies that have more complex geometry.
> For example, you can build concave shapes. When a child shape is added to
> the Compound shape, a transform consisting of a position and a quaternion is
> needed. This enables you to add child shapes at any position, rotated however
> you like inside the local coordinate system of the Compound shape.

This might be best suited for 2.0

**Comment by codynova** [comment]
I think this is actually pretty straightforward - I've used this before to create rigid player bodies for a game scene. We need to expose `addShape` on the `useBody` api. I can take a look soon.

**Comment by codynova** [comment]
We might also consider a more automated way of creating compound shapes with a new hook. Any suggestions what the api might look like? Maybe something similar to this?

```js
const halfExtents = new Vec3(1, 1, 1)
const shape = { type: 'Box', args: [ halfExtents ] }
const [ref, api] = useCompoundBody(() => ({ mass: 100 }, [
  { ...shape, position: new Vec3(2, 0, 0) },
  { ...shape, position: new Vec3(4, 0, 0) },
]))
```

**Comment by nevernotsean** [comment]
This seems pretty ideal, supplying an array of objects with the appropriate body arguments is intuitive enough. I would convert the `new Vec3(...)` to an array of numbers to keep the api consistent

**Comment by codynova** [comment]
>I would convert the new Vec3(...) to an array of numbers to keep the api consistent

Yep great point!

**Comment by codynova** [comment]
I have this mostly working locally but I really want to clean up the `cannon-es` types before I go any further as things are starting to feel a bit messy without the intellisense... I created a new PR in the `cannon-es` repo for discussion around the possibility of introducing TypeScript to `cannon-es`. The PR is sitting in closed status as it's not ready to be merged.

**Comment by codynova** [comment]
While working on this, I noticed that Cannon doesn't recompute the center of mass for compound bodies. This will need to be done manually in the worker each time a shape is added:
```js
function updateCenterOfMass (body) {
    const centerOfMass = new Vec3()
    body.shapeOffsets.forEach(offset => centerOfMass.vadd(offset, centerOfMass))
    centerOfMass.scale(1/body.shapes.length, centerOfMass)
    body.shapeOffsets.forEach(offset=> offset.vsub(centerOfMass, offset))
    const worldCenterOfMass = new Vec3()
    body.vectorToWorldFrame(centerOfMass, worldCenterOfMass )
    body.position.vadd(worldCenterOfMass, body.position)
};
```

taken from https://github.com/schteppe/cannon.js/issues/230#issuecomment-141810570

**Comment by codynova** [comment]
Exposed `Body.prototype.addShape` in `api.addShapes` #40

**Comment by codynova** [comment]
Added `useCompoundBody` hook and demo in #42:

```js
const [ref] = useCompoundBody(() => ({
    mass: 12,
    ...props,
    shapes: [
        { type: 'Box', position: [0, 0, 0], rotation: [0, 0, 0], args: [0.5, 0.5, 0.5] },
        { type: 'Sphere', position: [1, 0, 0], rotation: [0, 0, 0], args: [0.65] },
    ],
}))
```

**Comment by codynova** [comment]
Merged #42. `useCompoundBody` will be available on the next release.
